_posts/2021-02-16-message-passing.md

```diff
@@ -0,0 +1,209 @@
+the guarantees stop there. If more than one process sends messages to a common
+process, they can arrive in any order even when you "know" that one of the
+messages was sent first. For example, this sequence of events is legal and
+entirely possible:
+
+1. `A` sends message `1` to `B`
+1. `A` sends message `2` to `C`
+1. `C`, in response to message `2`, sends message `3` to `B`
+1. `B` receives message `3`
+1. `B` receives message `1`
+
+Luckily, global orders are rarely needed and are easy to impose yourself
+(outside distributed cases): just let all involved parties synchronize with a
+common process.
+
+### Sending messages
+
+Sending a message is straightforward: we try to find the process associated
+with the process identifier, and if one exists we insert the message into its
+signal queue.
+
+Messages are always copied before being inserted into the queue. As wasteful as
+this may sound it greatly reduces garbage collection (GC) latency as the GC
+never has to look beyond a single process, which is more important than sheer
+throughput for the kind of soft-realtime systems Erlang is designed to build.
+
+By default, messages are copied directly into the receiving process' heap but
+when this isn't possible (or desired -- see the [message_queue_data] flag) we
+allocate the message outside of the heap instead.
+
+Memory allocation makes such "off-heap" messages slightly more expensive but
+they're very neat for processes that receive a ton of messages. We don't need
+to interact with the receiver when copying the message -- only when adding it
+to the queue -- and since the only way a process can see a message is by
+matching them in a `receive` expression, the GC doesn't need to consider
+unmatched messages which further reduces latency.
+
+Sending messages to processes on other Erlang nodes works in the same way,
+albeit there's now a risk of messages being lost in transit. Messages are
+guaranteed to be delivered as long as the distribution link between the nodes
+is active, but it gets tricky when the link goes down.
+
+Using `monitor/2` on the remote process (or node) will tell you when this
+happens, acting as if the process died (with reason `noconnection`), but that
+doesn't always help: the link could've died _after_ the message was received
+and handled on the other end, all we know is that the link went down _before_
+we got any eventual response.
+
+As with everything else there's no free lunch, and you need to decide how your
+applications should handle [these scenarios].
+
+### Receiving messages
+
+One might guess that processes receive messages through `receive` expressions,
+but `receive` is a bit of a misnomer. As with all other signals the process
+continuously handles them in the background, moving received messages from the
+_signal_ queue to the _message_ queue.
+
+`receive` searches for matching messages in the message queue (in the order
+they arrived), or waits for new messages if none were found. Searching through
```

Comment:
Suggest to add queue: Searching through the message +queue+ rather than the signal queue means ... 
I don't really understand "doesn't have to contend with processes that send messages" please clarify or rephrase